### PR TITLE
Example usage of CREATE2 deployer proxy

### PR DESCRIPTION
**Pull request by saucepoint**
Providing an example of how to use the CREATE2 deployer proxy

### CODE REVIEW

File: src/reference/anvil/README.md [hunk 1]
```diff
@@ -37,10 +37,36 @@ anvil --port <PORT>
...
+    ))))
+);
+
+// deploying `Counter` with CREATE2 and salt
+(bool success, ) = CREATE2_DEPLOYER.call(
+    abi.encodePacked(salt, creationCode)
+);
+require(success, "Deployment failed");
+require(counterAddress.code.length > 0, "Contract not deployed");
+```
```
Comment (mds1) [inline src/reference/anvil/README.md:68]: I don't understand what this code snippet is for? From solidity scripts and tests, you should never need to manually encode this way if you want to use the create2 deploy. `new Counter{salt: mySalt}()` is all you need

Comment (saucepoint) [reply src/reference/anvil/README.md:68]: oh interesting, TIL!

Looks like `new Counter{salt: mySalt}()` works as expected in a script running against anvil, but doesnt seem to work in forge-test. The snippet I provided works in both

Comment (mds1) [reply src/reference/anvil/README.md:68]: It should work in `forge test` as long as the deployer is there locally. If it's not, I'd suggest "deploying" it with `vm.etch("7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe03601600081602082378035828234f58015156039578182fd5b8082525050506014600cf3")`, where that code is the deployed bytecode from https://etherscan.io/address/0x4e59b44847b379578588920ca78fbf26c0b4956c.

Comment (saucepoint) [reply src/reference/anvil/README.md:68]: isnt the deployer already available in `forge test` if the verbose snippet works? either way tried etching the bytecode and still getting smacked around 🤕

script (works): https://github.com/saucepoint/v4-template/blob/create2-test/script/Counter.s.sol#L23-L36

test (fails): https://github.com/saucepoint/v4-template/blob/create2-test/test/Counter.t.sol#L36-L48

Comment (mds1) [reply src/reference/anvil/README.md:68]: The test is failing from `HookAddressNotValid` though, which seems unrelated to the create2 deployer? (unless the address is being computed wrong and using the wrong deployer address)

Comment (saucepoint) [reply src/reference/anvil/README.md:68]: was mostly confused why the same code in `forge script` and `forge test` had different outcomes. Turns out calculating salts / precomputed addresses have different behaviors in script and test

---

Assume:
```solidity
    /// @notice Precompute a contract address deployed via CREATE2
    function computeAddress(address deployer, uint256 salt, bytes memory creationCode) public pure returns (address) {
        return address(
            uint160(uint256(keccak256(abi.encodePacked(bytes1(0xff), deployer, salt, keccak256(creationCode)))))
        );
    }
 ```

in forge script you need to provide `deployer = 0x4e59b44847b379578588920cA78FbF26c0B4956C`

in forge test you need to provide `deployer = address(this)` (or the pranking address)

---

willing to update the docs with my findings if you think this is worthwhile information

Comment (mds1) [reply src/reference/anvil/README.md:68]: Ah, yea this is a known issue and we should update tests to use the same create2 deployer as scripts, ref https://github.com/foundry-rs/foundry/issues/5529. cc @Evalir

**Review by muntahafayiz — APPROVED** [review]

File: src/reference/anvil/README.md [hunk 2]
```diff
@@ -37,10 +37,36 @@ anvil --port <PORT>
...
+
 #### Supported RPC Methods
 ##### Standard Methods
 The standard methods are based on [this](https://eth.wiki/json-rpc/API) reference.
```
Comment (muntahafayiz) [inline src/reference/anvil/README.md:72]: نسخة مكررة من #

File: src/reference/anvil/README.md (hunk 1 above)
Comment (mds1) [reply src/reference/anvil/README.md:68]: This is now resolved with the `always_use_create_2_factory` flag introduced in https://github.com/foundry-rs/foundry/pull/6656

**Comment by onbjerg** [comment]
Closing this as stale, feel free to re-open it if progress is made on it:)

**Pull request closed without merging**